uast/ann/folder.go

```diff
@@ -0,0 +1,95 @@
+package ann
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+// folder as in "analyzer of recursive data structures combining the
+// information in its nodes", not as in "directory".  This particular
+// folder traverses Rules in pre-order, generating their string
+// XPath-like descriptions.
+type folder struct {
```

Comment:
I would imagine `folder` as an interface implementing the `Fold()` method. Since this is an implementation, it should have a name descriptive of what it is actually doing.